internal/backend/workloadproxy/reconcile_utils.go

```diff
@@ -0,0 +1,283 @@
+// finds the first cluster with the given ID and returns its hosts and in-use port, also setting the in-use port for the
+// cluster.
+func (a *aliasToCluster) ActiveHostsPort(clusterID resource.ID) ([]string, port) {
+	val, ok := a.slc.FindBy(func(d *clusterData) bool { return d.clusterID == clusterID })
+	if !ok {
+		return nil, ""
+	}
+
+	for _, v := range a.aliasToPort.All() {
+		if v.F1 == val {
+			val.inUsePort = v.F2
+
+			return val.hosts, v.F2
+		}
+	}
+
+	return nil, ""
+}
+
+func (a *aliasToCluster) DropAlias(als alias) *clusterData {
+	removed, ok := a.aliasToPort.Delete(als)
+	if !ok {
+		return nil
+	}
+
+	clusterPtr := removed.F1
+
+	if clusterPtr.inUsePort == removed.F2 {
+		clusterPtr.inUsePort = ""
+	}
+
+	return clusterPtr
+}
+
+// All returns all the alias data. The function returns a triple: alias, it's
+// port and the cluster data.
+func (a *aliasToCluster) All() iter.Seq2[pair.Pair[alias, port], *clusterData] {
+	return func(yield func(pair.Pair[alias, port], *clusterData) bool) {
+		for als, v := range a.aliasToPort.All() {
+			ptr := v.F1
+			if ptr == nil {
+				continue
+			}
+
+			if !yield(pair.MakePair(als, v.F2), ptr) {
+				return
+			}
+		}
+	}
+}
+
+type clusterData struct {
+	clusterID resource.ID
+	inUsePort port
+	hosts     []string
+	aliases   []alias
+}
+
+func (a *clusterData) Equal(other *clusterData) bool { return a.clusterID == other.clusterID }
+
+// uniqueSlice is a data structure that ensures that the elements are unique
+// (using Equal(T) contract). It also allows for a callback to be called when the
+// element is removed.
+type uniqueSlice[E elem[E]] []E
+
+// Replace replaces the element in the slice. If the element doesn't exist, it's
+// added to the slice. The function returns the old element and a boolean
+// indicating if the element was replaced.
+func (s *uniqueSlice[E]) Replace(val E) (old E, replaced bool) {
+	idx := slices.IndexFunc(*s, func(e E) bool { return e.Equal(val) })
+	if idx != -1 {
+		prev := (*s)[idx]
+		(*s)[idx] = val
+
+		return prev, true
+	}
+
+	*s = append(*s, val)
+
+	return *new(E), false
+}
+
+// FindBy finds the element in the slice. If the element doesn't exist, the
+// function returns the zero value of the element and false.
+func (s *uniqueSlice[E]) FindBy(pred func(E) bool) (E, bool) {
+	if idx := slices.IndexFunc(*s, func(e E) bool { return pred(e) }); idx != -1 {
+		return (*s)[idx], true
+	}
+
+	return *new(E), false
+}
+
+// Remove removes the element from the slice.
+func (s *uniqueSlice[E]) Remove(val E) bool {
+	if idx := slices.IndexFunc(*s, func(e E) bool { return e == val }); idx != -1 {
+		*s = slices.Delete(*s, idx, idx+1)
+
+		return true
+	}
+
+	return false
+}
+
+type elem[E any] interface {
+	comparable
+	Equal(E) bool
+}
+
+// actionMap is a data structure that maps keys to values.
+type actionMap[K comparable, V any] struct {
+	m map[K]V
```

Comment:
I don't think it should be called `actionMap` anymore. As there are no actions here.

Reply:
And it can be actually replaced by a regular map

Reply:
👍